File: solutions/typescript/clock/1/clock.ts
class Clock {
  private _hours!: number;
  private _minutes!: number;

  constructor(hours: number, minutes: number = 0) {
    this.hours = hours;
    this.minutes = minutes;
  }

  public get hours() {
    return this._hours;
  }

  public set hours(value: number) {
    const nextHours = value % 24;

    if (nextHours < 0) {
      this._hours = 24 + nextHours;
    } else {
      this._hours = nextHours;
    }
  }

  public get minutes() {
    return this._minutes;
  }

  public set minutes(value: number) {
    const potentialHours = Math.floor(value / 60);
    const potentialMinutes = value % 60;

    this.hours += potentialHours;

    if (potentialMinutes < 0) {
      this._minutes = 60 + potentialMinutes;
    } else {
      this._minutes = potentialMinutes;
    }
  }

  public plus(minutes: number) {
    this.minutes += minutes;

    return this;
  }

  public minus(minutes: number) {
    this.minutes -= minutes;

    return this;
  }

  public toString() {
    return `${this.formatNumber(this.hours)}:${this.formatNumber(
      this.minutes
    )}`;
  }

  public equals(clock: Clock) {
    return this.minutes === clock.minutes && this.hours === clock.hours;
  }

  private formatNumber(n: number) {
    if (n < 10) {
      return `0${n}`;
    }

    return n;
  }
}

export default Clock;

File: solutions/csharp/clock/1/Clock.cs
public class Clock
{

    private const int HoursPerDay = 24;
    private const int MinutesPerHour = 60;
    private static readonly int MinutesPerDay = MinutesPerHour * HoursPerDay;

    private int _minutes;

    public Clock(int hours, int minutes)
    {
        _minutes = (minutes + (hours % HoursPerDay) * MinutesPerHour) % MinutesPerDay;

        while (_minutes < 0)
        {
            _minutes += MinutesPerDay;
        }
    }
    
    public Clock Add(int minutesToAdd) {
        _minutes += minutesToAdd;
        _minutes %= MinutesPerDay;
        
        return this;
    }

    public Clock Subtract(int minutesToSubtract)
    {
        _minutes -= minutesToSubtract;

        while (_minutes < 0)
        {
            _minutes += MinutesPerDay;
        }

        return this;
    }

    public override string ToString()
    {
        var hours = _minutes / MinutesPerHour;
        var minutes = _minutes % MinutesPerHour;
        
        return $"{hours:D2}:{minutes:D2}";   
    }

    public override bool Equals(object? obj)
    {
        if (obj is Clock clock)
        {
            return _minutes == clock._minutes;
        }

        return false;
    }

    public override int GetHashCode() => _minutes.GetHashCode();
}
